processing/src/main/java/org/apache/druid/query/aggregation/AggregatorFactory.java

```diff
@@ -211,22 +213,35 @@ public AggregatorFactory getMergingFactory(AggregatorFactory other) throws Aggre
   public abstract List<String> requiredFields();
 
   /**
-   * Get the type name of the intermediate type for this aggregator. This is the same as the type returned by
+   * Get the "intermediate" {@link ValueType} for this aggregator. This is the same as the type returned by
    * {@link #deserialize} and the type accepted by {@link #combine}. However, it is *not* necessarily the same type
    * returned by {@link #finalizeComputation}.
+   */
+  public abstract ValueType getType();
```

Comment:
thinking out loud. how about having a class such as `AggregatorOutputType` which can contain `combine` and `finalize` types. Aggregators just override one method which can be `AggregatorOutputType getType()`

Reply:
over time, more information can be added to this class rather than having methods in the interface. one example is complex type name.

Reply:
We already have `ColumnCapabilities` which is a richer but a bit more segment focused version of a thing that provides details about a given column, and #10277 expands that to also include complex type name, so I think if we are going to use a consolidated type we should consider using that since it is used pretty heavily throughout the engines to help determine the correct way to process columns (or a new interface that provides a subset of the functionality that `ColumnCapabilities` can extend as mentioned in another thread). I need to think a bit harder about what would be most useful for `RowSignature` to have on hand, which is currently the type used to serve similar functionality to higher levels of the engine what `ColumnCapabilities` provides at lower levels, before making a change like this though.

I'm not sure if it is useful to combine the intermediary and finalized types in one place, since callers should typically only need one or the other, depending on the caller, as mentioned in PR description and this thread https://github.com/apache/druid/pull/9638#discussion_r417758925. We currently haven't quite added enough information through for most callers to be able to provide this context quite yet though, so it also needs a bit further thought.